heartbeat/skima/core.go

```diff
@@ -0,0 +1,168 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package skima
+
+import (
+	"strings"
+
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"fmt"
+
+	"sort"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// Is creates a named IsDef with the given checker.
+func Is(name string, checker Checker) IsDef {
+	return IsDef{name: name, checker: checker}
+}
+
+// Optional wraps an IsDef to mark the field's presence as optional.
+func Optional(id IsDef) IsDef {
+	id.name = "optional " + id.name
+	id.optional = true
+	return id
+}
+
+// Map is the type used to define schema definitions for Schema.
+type Map map[string]interface{}
+
+// SchemaValidator is the result of Schema and is run against the map you'd like to test.
+type SchemaValidator func(common.MapStr) SchemaValidationResults
+
+// Compose combines multiple SchemaValidators into a single one.
+func Compose(validators ...SchemaValidator) SchemaValidator {
+	return func(actual common.MapStr) SchemaValidationResults {
+		results := make([]SchemaValidationResults, len(validators))
+		for idx, validator := range validators {
+			results[idx] = validator(actual)
+		}
+
+		combined := SchemaValidationResults{}
+		for _, r := range results {
+			r.eachResult(func(path string, vr ValueResult) {
+				combined.recordResult(path, vr)
+			})
+		}
+		return combined
+	}
+}
+
+// Strict is used when you want any unspecified keys that are encountered to be considered errors.
+func Strict(laxValidator SchemaValidator) SchemaValidator {
+	return func(actual common.MapStr) SchemaValidationResults {
+		validatedResults := laxValidator(actual)
+
+		// The inner workings of this are a little weird
+		// We use a hash of dotted paths to track the results
+		// We can check if a key had a test associated with it by looking up the laxValidator
+		// result data
+		// What's trickier is intermediate maps, maps don't usually have explicit tests, they usually just have
+		// their properties tested.
+		// This method counts an intermediate map as tested if a subkey is tested.
+		// Since the datastructure we have to search is a flattened hashmap of the original map we take that hashmap
+		// and turn it into a sorted string array, then do a binary prefix search to determine if a subkey was tested.
+		// It's a little weird, but is fairly efficient. We could stop using the flattened map as a datastructure, but
+		// that would add complexity elsewhere. Probably a good refactor at some point, but not worth it now.
+		validatedPaths := sort.StringSlice{}
```

Comment:
while it works a more idiomatic use of sort package is to use it only for sorting, not for it's types:

e.g.
```
validatedPaths := make([]string, len(validatedResults))
for k := range validatedResults {
```